File: poo/java/HerancaAbstrata/Conta.java
/*
 *  Programa escrito com propósito didático apenas.
 *  @author Nathan Cruz <nathan.cruz@unifesp.br>
 *  @created 04/04/14
 *  R.A.: 86492
 */

public abstract class Conta {

    /* Atributos */
    protected double saldo;
    protected int numero;

    /* Construtor
    public Conta(int numero) {
      this.saldo = 0.0;
      this.numero = numero; 
    }*/

    public void credita(double quantia) {
        if (quantia > 0){
            this.saldo += quantia;
        }
        else{
            System.err.println("Não se pode depositar uma quantia negativa");
        }
    }

    public void debita(double quantia) {
        if (quantia>0){
            this.saldo -= quantia;
        }
        else{
            System.err.println("Não se pode sacar uma quantia negativa");
        }
    }

    public void atualiza(double taxa) {
        this.saldo += this.saldo * taxa; 
    }

    /* Getters */
    public double getSaldo(){
        return saldo;
    }
    public double getNumero(){
        return numero;
    }

    public void imprime(){
        System.out.println("Conta " + numero + ": " +
                           "saldo = " + saldo );
    }
}

File: poo/python/HerancaAbstrata/Conta.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-

'''
@author Nathan Cruz <nathan.cruz@unifesp.br>
@created 13/04/14
'''

from __future__ import unicode_literals
from __future__ import print_function

# Modulo contendo metaclass e outras coisas uteis para classes
# abstratas
import abc


class Conta(object):

    # Transforma a classe numa classe abstrata
    __metaclass__ = abc.ABCMeta


    # Atributos
    __saldo = 0.0
    __numero = None


    # Faz um deposito na conta
    def credita(self, quantia):
        if quantia > 0:
            self.__saldo += quantia
        else:
            raise ValueError('{0} {1}'.format('Nao se pode sacar',
                                              'zero ou menos'))


    # Faz um saque da conta
    def debita(self, quantia):
        if quantia <= self.__saldo:
            if quantia <= 0:
                raise ValueError('{0} {1}'.format('Nao se pode sacar',
                                                  'zero ou menos'))
            self.__saldo -= quantia
        else:
            raise ValueError('{0} {1}'.format('Nao se pode sacar',
                                              'mais do que se tem'))


    # Getters
    def getSaldo(self):
        return self.__saldo


    def getNumero(self):
        return self.__numero


    # Representacao da classe
    def imprime(self):
        print('Conta\n\tIdentificador: {0}\n\tSaldo: R${1:.2f}'.
              format(self.__numero, self.__saldo))
